docs/rfcs/1011-non-blocking-invite-transport.md

```diff
@@ -0,0 +1,1238 @@
+- created-on timestamp
+- created-by author
+- claimer id:
+  * email address (for user invitation)
+  * or user id (for device or shamir invitation)
+
+Mutable state:
+- state: `PENDING` | `CANCELLED` | `COMPLETED`
+
+Getters:
+- get greeters
+  * for user invitation: all administrators
+  * for device invitation: same as claimer id
+  * for shamir invitation: all recipients
+- get channel for greeter
+
+Methods:
+- create invitation
+- cancel invitation
+- complete invitation
+
+
+### Channel
+
+A channel is dedicated to the verification of the claimer identity by a specific greeter.
+
+A channel is identified by:
+- organization id
+- invitation token
+- greeter id
+
+Mutable properties:
+- attempts ids: list of attempts
+   * Invariant: most recent attempt is active, all the rest is cancelled
+
+Getters:
+- get active attempt
+
+Methods:
+- cancel active attempt (and create a new one)
+
+
+### Attempt
+
+An attempt is the steps followed by the peers of a given channel when trying to establish a secure communication channel.
+
+An attempt is identified by:
+- organization id
+- invitation token
+- greeter id
+- attempt id
+
+Mutable properties:
+- claimer joined: `NotSet` | `Timestamp`
+- greeter joined: `NotSet` | `Timestamp`
+- cancelled reason: `NotSet` | (`CLAIMER` | `GREETER`, `Reason`, `Timestamp`)
+
+Inferred properties:
+- state: `IDLE` | `HALF_JOINED` | `FULLY_JOINED` | `CANCELLED`
+- is active: `True` if it is either `IDLE`, `HALF_JOINED` or `FULLY_JOINED`
+
+Getters:
+- get steps
+
+Methods:
+- claimer join attempt
+- greeter join attempt
+- claimer cancel attempt
+- greeter cancel attempt
+
+
+### Step
+
+A step corresponds to the exchange of two messages between the peers in the context of an attempt.
+
+A step is identified by:
+- organization id
+- invitation token
+- greeter id
+- attempt id
+- step count
+
+Mutable properties:
+- claimer data: `NotSet` | `ClaimerStep`
+- greeter data: `NotSet` | `GreeterStep`
+
+Inferred properties:
+- state: `PENDING` | `HALF_PENDING` | `COMPLETED`
+
+Methods:
+- set claimer data
+- set greeter data
+
+
+## Design
+
+### New commands
+
+7 commands are added in total:
+- 3 new invited commands
+    - [`invite_claimer_start_attempt`](#invite_claimer_start_attempt-command)
+    - [`invite_claimer_cancel_attempt`](#invite_claimer_cancel_attempt-command)
+    - [`invite_claimer_step`](#invite_claimer_step-command)
+- 4 new authenticated commands
+    - [`invite_greeter_start_attempt`](#invite_greeter_start_attempt-command)
+    - [`invite_greeter_cancel_attempt`](#invite_greeter_cancel_attempt-command)
+    - [`invite_greeter_step`](#invite_greeter_step-command)
+    - [`invite_complete`](#invite_complete-command)
+
+#### `invite_greeter_start_attempt` command
+
+Two new commands are introduced to start an attempt, which returns an attempt ID used in the subsequent queries. This attempt ID is the same for the greeter and the claimer. For the greeter, an authenticated command is exposed.
```

Comment:
```suggestion
Two new commands are introduced to start an attempt, which returns an attempt ID used in the subsequent queries. This attempt ID is the same for the greeter and the claimer and is a UUID that uniquely identify the attempt. For the greeter, an authenticated command is exposed.
```

See my remarks below, I think it's much better to only use the attempt ID to identify the attempt that rely on the couple token+attempt ID.

On top of that I don't think it's interesting to have the attempt ID being an index (so first attempt has ID 0, next one 1 etc.) as we don't care of the number of previous attempts made, and handle counter is always more complicated to implement than plain simple UUID.

Reply:
> See my remarks below, I think it's much better to only use the attempt ID to identify the attempt that rely on the couple token+attempt ID.

I don't agree it's better, I would argue it's worse. Your main argument is that this makes room for inconsistency, as pointed in this comment:

> Same as above: token+greeter+attemptID makes rooms for inconsistency, we should only have a attempt field instead

However the attempt is always part of a specific invitation and a specific channel, which means we do need to make sure that those are valid in order to deal with the attempt. More specifically, we need to make sure that:
- the invitation is not cancelled
- the invitation is not completed
- the greeter is allowed (not revoked)

The corresponding pseudo-code is this:
```python
invitation = organization.get_invitation(invitation_token)
assert invitation.is_active()
channel = invitation.get_channel(greeter_id)
assert channel.is_active()
attempt = channel.get_attempt(attempt_id)
assert attempt.is_active()
```

I don't see room for inconsistency: the attempt is simply unique in the context of a specific channel. The fact that it is identified by an UUID and then unique globally is merely an implementation detail. It's actually already the same for invitation token and organization ID: invitation token are expected to be unique in a given organization, although they are usually unique globally.

In my opinion, there is nothing to gain by having the attempt ID global to an organization: it serves no purpose other than removing the invitation token from the commands, although the server implementation would still have to get the corresponding invitation in order to perform the proper checks.

@mmmarcos @FirelightFlagboy @AureliaDolo Do you have an opinion about it?

Reply:
> However the attempt is always part of a specific invitation and a specific channel, which means we do need to make sure that those are valid in order to deal with the attempt. More specifically, we need to make sure that:
>
>    the invitation is not cancelled
>    the invitation is not completed
>    the greeter is allowed (not revoked)

I agree

> The corresponding pseudo-code is this:

I think it's more practical to use pseud-SQL to illustrate here.

With UUID for attempt ID and channel ID:
```sql
SELECT
    invitation.is_active as invitation_is_active,
    channel.is_active as channel_is_active,
    attempt.is_active as attempt_is_active
FROM
    attempt
    LEFT JOIN channel ON attempt.channel = channel._id
    LEFT JOIN invitation ON channel.invitation = invitation._id
WHERE attempt.attempt_id = $?;
```

VS with counters for attempt ID and channel ID:
```sql
SELECT
    invitation.is_active as invitation_is_active,
    channel.is_active as channel_is_active,
    attempt.is_active as attempt_is_active
FROM
    attempt
    LEFT JOIN channel ON attempt.channel = channel._id
    LEFT JOIN invitation ON channel.invitation = invitation._id
WHERE
    invitation.invitation_id = $? AND
    channel.channel_id = $? AND
    attempt.attempt_id = $?
```

This is just needlessly more complicated.

On top of that we have to consider the creation of a new channel & attempt: with UUID it's trivial, but with counter-based IDs we need to first request the database to get the next value or to compute the value in the request then use a returning clause to get back the value in our code.

Finally while having a counter for the attempt makes sens given it represent the attempts a client has done, the rational is much weaker for the channel (where the counter represent the order in which greeters has accepted to do the invitation).

Of course we could avoid using counter-based ID and rely on UUID-but-not-globally-unique (or using the internal ID for the row in SQL) for attempt ID & channel ID, but at this point we are just pretending the ID is not globally unique ;-)

> I don't see room for inconsistency

You are correct, I was assuming channel ID & attempt ID where UUID AND where provided (along with invitation ID)  in the request, in which case only attempt ID is used for the query, then invitation ID & channel ID must be checked against the query output (hence the possible inconsistency).

> The fact that it is identified by a UUID and that it's unique globally is merely an implementation detail
>
> In my opinion, there is nothing to gain by having the attempt ID global to an organization
>
> On the contrary, keeping proper scoping for the different IDs makes it explicit that the attempt command only makes sense in the context of a valid invitation and channel

I think those three parts summary well the situation ^^
I am in favor of using UUID for pragmatic reasons:
- it simplifies the implementation (sql queries + errors handling)
- the only thing we sacrifice is being able to order channels/attempt but we don't need it (I would even say adding order brings implicit rules to the database, for instance doing `SELECT .. FROM attempt WHERE is_active = TRUE` is implicitly expected to return the attempt with the higher counter ID)

In the end the question is how much do we value "different IDs makes it explicit that the attempt command only makes sense in the context of a valid invitation and channel" ? If you think it is of high important then go with counter, if not got with UUID (and if you need outside opinion:  I'm strongly in favor of UUID, especially given it's cheep to add a comment next in the invitation commands to explicit the relashionship between attempt/channel/invitation).

Reply:
**Pavé Cesar !**

If I understand the discussion correctly, I would resume it to the following question:

> Should we use an UUID or a counter for `attempt_id` ?

## My observations

### Either UUID or counter

- The clients/server would still need to check the validity of attempt/channel/invitation.

### With UUID

- Could be used to identify the couple `attempt/channel/invitation` with it's sole value.
- Prevent to sort the attempts.

### With counter

- Require additional information to be provided to identify the couple `attempt/channel/invitation`.
- Could sort the attempts.

## My Opinions

### Regarding the information to be provided to identify the couple `attempt/channel/invitation`

Requiring to pass `attempt_id`, `channel_id` and `invitation_id`: 

I assume would increase the bookkeeping of those value during the invitation process.
But that information that we already have on hand and would ensure we are referencing the correct invitation process.

Identity the couple with only `attempt_id`:

That only possible with an ID with a high entropy (like UUID).

### Validating the couple

Either way, we would have to use the all the ids either by using the ones provided or by fetching them in the database.

From the 2 examples provided by @touilleMan, I'm fine with both.

## Conclusion

If I want simplicity, I would choose the UUID for the `attempt_id` given it would identify the couple `attempt/channel/invitation` with a single value.

If I want to be stricter, I would require use additional ids alongside `attempt_id`, like:

- `attempt_id + channel_id`
- `attempt_id + invitation_id`
- `attempt_id + channel_id + invitation_id`

> [!NOTE]
> With the stricter way, using an UUID or a counter is an implementation details, 
> given we would define the invitation with an additional id that is an UUID.

> I would be in a more peace of mind state with the stricter approach.

## Side dish: What about ULID over UUID ?

It as the advantage of both a counter and an UUID:

- It's unique (but it's partialy generated with a timestamp) with an entropy of 128bit.
- Is sortable.
- Is compatible with UUID.

<https://github.com/ulid/spec>

Reply:
I don't see an actual "debate" regarding UUIDs here:

- Identifying an attempt with a UUID seems like the right implementation choice. The RFC does not state the contrary since it does not make any assumptions on that (except maybe for the choice of fields in the requests, see my answer below).
- There is maybe a misunderstanding regarding phrase like ("An attempt is identified by: ..."). Those will definitely be required fields (foreign keys) and may form and alternate key in the attempt schema. However, the schema will most probably be implemented with a UUID surrogate key (`attempt_id`) as this is a common practice with many benefits over natural/composite keys.

On the other hand, when discussing what is used in the requests (see https://github.com/Scille/parsec-cloud/pull/7662/files#r1680380598), we are not talking about an "implementation detail" anymore, but an API design choice. 
IMO, a request parameter should be mandatory if it is a _sine qua non_ condition to fulfill the request (and optional if it allows to tune the response). I don't think this is the case for the invitation `token` parameter if the goal is to retrieve a specific `attempt` so it feels weird to having it in the request (as one would not put `name` or `email` together with a `user_id` to retrieve a specific user). So unless the invitation `token` serves another purpose, it seems redundant to me and this would be the main reason to avoid it (regardless if this makes one response status more or less to manage, which I think is not the essential to the discussion).

In any case, having retrieved the right attempt does not avoid the need to perform the validation steps described by @vxgmichel in this RFC.

Reply:
ccf6d568d950206e8b632bc2469462e7910f790c